File: typescript/src/03-observer/after/order-status-publisher.ts
import type { IOrderStatusSubscriber } from './order-status-subscriber.js';
import type { Order } from '../../shared/models.js';

// Publisher holds a dynamic list of subscribers and broadcasts one event.
//    It has no idea who is listening.
export class OrderStatusPublisher {
  private readonly subscribers: IOrderStatusSubscriber[] = [];

  subscribe(subscriber: IOrderStatusSubscriber): void {
    this.subscribers.push(subscriber);
  }

  setStatus(order: Order, status: string): void {
    order.status = status;
    console.log(`  📢 Publisher: Order ${order.id} status -> ${status}`);

    // Fire-and-forget: we do NOT await the subscriber promises, so the main flow
    // returns immediately. This is the TS analog of C#'s Task.Run - DEMO ONLY.
    // Unawaited promise rejections are swallowed (see README); production uses a
    // proper queue (BullMQ / SQS / Kafka) with real error handling.
    for (const subscriber of this.subscribers) {
      void subscriber.onStatusChanged(order);
    }

    console.log('  [After] Publisher main thread free - subscribers running in background');
  }
}

File: csharp/DesiArchitect.DesignPatterns/03_Observer/After/OrderStatusPublisher.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Observer.After;

public class OrderStatusPublisher
{
    private readonly List<IOrderStatusSubscriber> _subscribers = new();

    public void Subscribe(IOrderStatusSubscriber subscriber) => _subscribers.Add(subscriber);

    public void SetStatus(Order order, string status)
    {
        order.Status = status;
        Console.WriteLine($"  📢 Publisher: Order {order.Id} status -> {status}");
        foreach (var subscriber in _subscribers)
            Task.Run(() => subscriber.OnStatusChanged(order));
        Console.WriteLine("  [After] Publisher main thread free - subscribers running in background");
    }
}
